# **Etapa 2: Análisis de datos exploratorio**


En esta práctica realizaremos un poco de análisis de datos exploratorio en la búsqueda de patrones en los datos y así determinar que valores deberíamos usar para realizar predicciones en un futuro.

Veremos si podemos predecir la probabilidad de éxito basandonos en ciertas categorías.



## Objectivos

Realizar un análisis exploratorio de los datos y obtener nuestras primeras conclusiones.


***


## Importación de librerias

In [1]:
import pandas as pd
import numpy as np

### Análisis de datos.


Carga el archivo `.csv` generado en la práctica anterior. Recuerda que primero deberás subir el archivo a google collab.

In [65]:
df=pd.read_csv("dataset_part_1.csv")
#df.head(10)
df[:10]

FlightNumber        Date BoosterVersion  PayloadMass  \
0             1  2010-06-04       Falcon 9   8191.07911   
1             2  2012-05-22       Falcon 9    525.00000   
2             3  2013-03-01       Falcon 9    677.00000   
3             4  2013-09-29       Falcon 9    500.00000   
4             5  2013-12-03       Falcon 9   3170.00000   
5             6  2014-01-06       Falcon 9   3325.00000   
6             7  2014-04-18       Falcon 9   2296.00000   
7             8  2014-07-14       Falcon 9   1316.00000   
8             9  2014-08-05       Falcon 9   4535.00000   
9            10  2014-09-07       Falcon 9   4428.00000   

                     Payload Orbit    LaunchSite LandingOutcome  Flights  \
0  Dragon Qualification Unit   LEO  CCSFS SLC 40      None None        1   
1         COTS Demo Flight 2   LEO  CCSFS SLC 40      None None        1   
2                      CRS-2   ISS  CCSFS SLC 40      None None        1   
3                   CASSIOPE    PO   VAFB SLC 4E    False Ocean        1   
4                      SES-8   GTO  CCSFS SLC 40      None None        1   
5                  Thaicom 6   GTO  CCSFS SLC 40      None None        1   
6                      CRS-3   ISS  CCSFS SLC 40     True Ocean        1   
7             Orbcomm-OG2-M1   LEO  CCSFS SLC 40     True Ocean        1   
8                  AsiaSat 8   GTO  CCSFS SLC 40      None None        1   
9                  AsiaSat 6   GTO  CCSFS SLC 40      None None        1   

   GridFins  Reused   Legs LandingPad  Block   Customers Mission_Outcome  \
0     False   False  False        NaN    1.0      SpaceX            True   
1     False   False  False        NaN    1.0  NASA(COTS)            True   
2     False   False  False        NaN    1.0  NASA (CRS)            True   
3     False   False  False        NaN    1.0         MDA            True   
4     False   False  False        NaN    1.0         SES            True   
5     False   False  False        NaN    1.0     Thaicom            True   
6     False   False   True        NaN    1.0  NASA (CRS)            True   
7     False   False   True        NaN    1.0     Orbcomm            True   
8     False   False  False        NaN    1.0     AsiaSat            True   
9     False   False  False        NaN    1.0     AsiaSat            True   

   ReusedCount Serial   Longitude   Latitude  
0            0  B0003  -80.577366  28.561857  
1            0  B0005  -80.577366  28.561857  
2            0  B0007  -80.577366  28.561857  
3            0  B1003 -120.610829  34.632093  
4            0  B1004  -80.577366  28.561857  
5            0  B1005  -80.577366  28.561857  
6            0  B1006  -80.577366  28.561857  
7            0  B1007  -80.577366  28.561857  
8            0  B1008  -80.577366  28.561857  
9            0  B1011  -80.577366  28.561857

### Tarea 1: Calcula el número de lanzamientos en cada plataforma.

Los datos contienen diferentes plataformas como:


*   El complejo de lanzamiento de Cabo Cañaveral (VAFB SLC 4E)
*   La base de las fuerzas aéreas de Vanderberg (SLC-4E)
* El centro de lanzamientos Kennedy (KSC LC 39A).
Veamos cual de ellas es mas popular:





*Utilizando el método `value_counts()` en la columna `LaunchSite` determina el número de lanzamientos en cada plataforma.*

In [4]:
#INSERTA AQUÍ TU CÓDIGO
df.LaunchSite.value_counts()

CCSFS SLC 40    93
KSC LC 39A      49
VAFB SLC 4E     26
Name: LaunchSite, dtype: int64

### Tarea 2: Calcula el número de lanzamientos para cada órbita.


Cada lanzamiento tiene como objetivo una órbita diferente, estas son las más comunes:


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/api/Images/Orbits.png)


*Utilizando el método `value_counts()` en la columna `Orbit` determina el número de lanzamientos en cada órbita.*



In [5]:
#INSERTA AQUÍ TU CÓDIGO
df.Orbit.value_counts()

VLEO     54
ISS      32
GTO      31
LEO      14
PO       13
SSO      11
MEO       5
GEO       2
TLI       2
ES-L1     1
HEO       1
SO        1
Name: Orbit, dtype: int64

### Tarea 3: Calcula el número de diferentes resultados de aterrizaje, distingue los exitosos de los fallidos.


*Utilizando el método `value_counts()` en la columna `LandingOutcome` determina el número de lanzamientos en cada órbita. Asigna el resultado a la varibale `landing_outcomes`.*


In [6]:
#INSERTA AQUÍ TU CÓDIGO
landing_outcomes = df.LandingOutcome.value_counts()
landing_outcomes

True ASDS      109
True RTLS       23
None None       19
False ASDS       7
True Ocean       5
False Ocean      2
None ASDS        2
False RTLS       1
Name: LandingOutcome, dtype: int64

Observemos que la variable `LandingOutcome` esta separada en 2 partes.

La primera toma valores `True`,`False` y `None`.
*   `True` implica que el cohete aterrizó con éxito.
*   `False` implica que el cohete no aterrizó con éxito
* `None` implica que el cohete no inició el aterrizaje.

La segunda parte toma valores `ASDS`,`Ocean`, `RTLS` y `None`.
* `ASDS` implica que el cohete tenía intención de aterrizar sobre un portaaviones.
* `Ocean` implica que el cohete tenía intención de aterrizar en el oceano.
* `RTLS` implica que el cohete tenía intención de aterrizar sobre una plataforma terrestre .
* `None` implica que el cohete no tenía intención de aterrizar.




Vamos a intentar ver que resultado ha tenido cada cohete en las diferentes órbitas.
*Crea un dataframe llamado `df2` con solamente las columnas `'Orbit','LandingOutcome'`. Aplica a este dataframe una agrupación por `'Orbit','LandingOutcome'` mediante `df2.groupby([Columnas])` y añade el atributo `.size()` para ver el número de resultados en cada categoría.

In [7]:
#INSERTA AQUÍ TU CÓDIGO
columnas_deseadas = ['Orbit','LandingOutcome']
df2 = df[columnas_deseadas]
df2.groupby(columnas_deseadas).size()

Orbit  LandingOutcome
ES-L1  True Ocean         1
GEO    True ASDS          2
GTO    False ASDS         1
       None ASDS          1
       None None         11
       True ASDS         17
       True Ocean         1
HEO    True ASDS          1
ISS    False ASDS         2
       False Ocean        1
       False RTLS         1
       None ASDS          1
       None None          3
       True ASDS         16
       True Ocean         1
       True RTLS          7
LEO    None None          2
       True ASDS          4
       True Ocean         1
       True RTLS          7
MEO    None None          1
       True ASDS          4
PO     False ASDS         1
       False Ocean        1
       None None          1
       True ASDS          8
       True Ocean         1
       True RTLS          1
SO     None None          1
SSO    True ASDS          4
       True RTLS          7
TLI    True ASDS          2
VLEO   False ASDS         3
       True ASDS         51
dtype: int64

En el resultado del anterior ejercicio podemos observar los diferentes resultados para cada órbita, que pese a ser detallados son confusos. Veamos como solucionar esto.


### Tarea 4: Crea una columna Class utilizando LandingOutcome


Previamente hemos guardado en `landing_outcomes` el recuento de los diferentes aterrizajes:

In [8]:
landing_outcomes

True ASDS      109
True RTLS       23
None None       19
False ASDS       7
True Ocean       5
False Ocean      2
None ASDS        2
False RTLS       1
Name: LandingOutcome, dtype: int64

*La variable anterior es un diccionario, utilizandola crea un conjunto con los diferentes casos donde el cohete no aterrizó satisfactoriamente. Llama a este conjunto `bad_outcomes`*

In [27]:
#INSERTA AQUÍ TU CÓDIGO
bad_outcomes = []
for i in landing_outcomes.keys():
    if "False"  in i or "None" in i:
        bad_outcomes.append(i)
bad_outcomes = set(bad_outcomes) # me la da en lista, la convierto en conjunto

Creemonos ahora una lista llamada `landing_class` que tomará valores 0 y 1 según si el resultado es satisfactorio o no.
* *Recorre para cada celda la columna `LandingOutcome`.*
* *Comprueba si el valor pertenece al conjunto `bad_outcomes`. Si pertenece adjunta un 0 a la lista `landing_class`, si no, adjunta un 1.*

In [28]:
#INSERTA AQUÍ TU CÓDIGO
landing_class = []
for classes in df.LandingOutcome:
    if classes in bad_outcomes:
        landing_class.append(0)
    else:
        landing_class.append(1)
df["Class"] = landing_class
df

FlightNumber        Date BoosterVersion  PayloadMass  \
0               1  2010-06-04       Falcon 9   8191.07911   
1               2  2012-05-22       Falcon 9    525.00000   
2               3  2013-03-01       Falcon 9    677.00000   
3               4  2013-09-29       Falcon 9    500.00000   
4               5  2013-12-03       Falcon 9   3170.00000   
..            ...         ...            ...          ...   
163           164  2022-08-28       Falcon 9  13260.00000   
164           165  2022-08-31       Falcon 9  13260.00000   
165           166  2022-09-17       Falcon 9  13260.00000   
166           167  2022-09-24       Falcon 9  13260.00000   
167           168  2022-10-05       Falcon 9   8191.07911   

                       Payload Orbit    LaunchSite LandingOutcome  Flights  \
0    Dragon Qualification Unit   LEO  CCSFS SLC 40      None None        1   
1           COTS Demo Flight 2   LEO  CCSFS SLC 40      None None        1   
2                        CRS-2   ISS  CCSFS SLC 40      None None        1   
3                     CASSIOPE    PO   VAFB SLC 4E    False Ocean        1   
4                        SES-8   GTO  CCSFS SLC 40      None None        1   
..                         ...   ...           ...            ...      ...   
163       Starlink 4-23 (v1.5)  VLEO    KSC LC 39A      True ASDS        2   
164        Starlink 3-4 (v1.5)  VLEO   VAFB SLC 4E      True ASDS        7   
165       Starlink 4-34 (v1.5)  VLEO  CCSFS SLC 40      True ASDS        6   
166       Starlink 4-35 (v1.5)  VLEO  CCSFS SLC 40      True ASDS        4   
167                     Crew-5   ISS    KSC LC 39A      True ASDS        1   

     GridFins  ...   Legs                LandingPad Block      Customers  \
0       False  ...  False                       NaN   1.0         SpaceX   
1       False  ...  False                       NaN   1.0     NASA(COTS)   
2       False  ...  False                       NaN   1.0     NASA (CRS)   
3       False  ...  False                       NaN   1.0            MDA   
4       False  ...  False                       NaN   1.0            SES   
..        ...  ...    ...                       ...   ...            ...   
163      True  ...   True  5e9e3033383ecb075134e7cd   5.0         SpaceX   
164      True  ...   True  5e9e3032383ecb6bb234e7ca   5.0         SpaceX   
165      True  ...   True  5e9e3033383ecbb9e534e7cc   5.0         SpaceX   
166      True  ...   True  5e9e3033383ecbb9e534e7cc   5.0         SpaceX   
167      True  ...   True  5e9e3033383ecbb9e534e7cc   5.0  NASA (CCtCap)   

    Mission_Outcome ReusedCount  Serial   Longitude   Latitude  Class  
0              True           0   B0003  -80.577366  28.561857      0  
1              True           0   B0005  -80.577366  28.561857      0  
2              True           0   B0007  -80.577366  28.561857      0  
3              True           0   B1003 -120.610829  34.632093      0  
4              True           0   B1004  -80.577366  28.561857      0  
..              ...         ...     ...         ...        ...    ...  
163            True           1   B1069  -80.603956  28.608058      1  
164            True           6   B1063 -120.610829  34.632093      1  
165            True           5   B1067  -80.577366  28.561857      1  
166            True           0   B1072  -80.577366  28.561857      1  
167            True           0   B1077  -80.603956  28.608058      1  

[168 rows x 21 columns]

*Añade esta lista al dataframe como una columna con nombre `Class`*

In [30]:
#INSERTA AQUÍ TU CÓDIGO
df.Class.unique()

array([0, 1])

## Utilizando la columna Class.

*Calcula la media de la columna `class`. Razona que significa este valor.*



In [31]:
#INSERTA AQUÍ TU CÓDIGO
df.Class.mean()

0.8154761904761905

Significa que tenemos un 81.5% de valores 1 y un 18.5% de valores 0. Lo comprobamos con la cuenta de cada valor:

In [32]:
df.Class.value_counts()

1    137
0     31
Name: Class, dtype: int64

31 de 0s, 137 de 1s, de los cuales 31 de 168 son 0s, un 18.5%.

Al igual que antes, podemos agrupar los lanzamientos por la órbita.
*Crea df3 donde incluyes las columnas `Orbit` y `Class`. Agrupa por `Orbit` y calcula la media.*

In [57]:
#INSERTA AQUÍ TU CÓDIGO
df3 = df.loc[:,["Orbit", "Class"]].groupby(df.Orbit)
df3.mean()

<ipython-input-57-4f2e7b221b74>:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df3.mean()


Class
Orbit          
ES-L1  1.000000
GEO    1.000000
GTO    0.580645
HEO    1.000000
ISS    0.750000
LEO    0.857143
MEO    0.800000
PO     0.769231
SO     0.000000
SSO    1.000000
TLI    1.000000
VLEO   0.944444

Observemos que más de una órbita tiene una probabilidad de éxito 1. Entonces ¿Si lanzamos un cohete a esas órbitas 100% aterrizará correctamente?

Tendremos mayores probabilidades si las lanzamos a esas órbitas, sí.

Mostremos ahora el número de lanzamientos éxitosos y fallidos en cada órbita.
*Aplica sobre df3 un método `groupby()` donde agrupes por las columnas `Orbit`y `Class`. Aplica el método `size()` para ralizar el conteo.

In [61]:
#INSERTA AQUÍ TU CÓDIGO
df3 = df.loc[:,["Orbit", "Class"]].groupby(["Orbit", "Class"])
df3.size()


Orbit  Class
ES-L1  1         1
GEO    1         2
GTO    0        13
       1        18
HEO    1         1
ISS    0         8
       1        24
LEO    0         2
       1        12
MEO    0         1
       1         4
PO     0         3
       1        10
SO     0         1
SSO    1        11
TLI    1         2
VLEO   0         3
       1        51
dtype: int64

Podemos observar que algunas de nuestras órbitas solamente tienen un lanzamiento y este exitoso (otras solamente uno fallido). En estos casos la muestra es demasiado pequeña como para asegurar la probabilidad de éxito solamente basandonos en la órbita. Por tanto, este no parece ser el méjor predictor de la probabilidad de éxito.

Retomemos las variables analizadas anteriormente como el LaunchSite.
*Crea df4 con las columnas `Class` y `LaunchSite`. Aplica un groupby y calcula la probabilidad de éxito para cada plataforma. Observa el número total de éxitos y fracasos en cada plataforma. Razona si sería un mejor predictor que la órbita.*

In [63]:
#INSERTA AQUÍ TU CÓDIGO
df4 = df.loc[:,["LaunchSite", "Class"]]
df4

LaunchSite  Class
0    CCSFS SLC 40      0
1    CCSFS SLC 40      0
2    CCSFS SLC 40      0
3     VAFB SLC 4E      0
4    CCSFS SLC 40      0
..            ...    ...
163    KSC LC 39A      1
164   VAFB SLC 4E      1
165  CCSFS SLC 40      1
166  CCSFS SLC 40      1
167    KSC LC 39A      1

[168 rows x 2 columns]

*Exporta los datos en un archivo llamado `dataset_part_2.csv` con `index=False`.
Descarga el archivo porque será necesario en la siguiente práctica.*


In [69]:
dff = df4.to_csv("dataset_part2_2.csv", index = False)